core/src/main/scala/kafka/coordinator/group/DelayedJoin.scala

```diff
@@ -33,11 +33,15 @@ import scala.math.{max, min}
  */
 private[group] class DelayedJoin(coordinator: GroupCoordinator,
                                  group: GroupMetadata,
-                                 rebalanceTimeout: Long) extends DelayedOperation(rebalanceTimeout, Some(group.lock)) {
+                                 rebalanceTimeout: Long) extends DelayedOperation(rebalanceTimeout, None) {
 
-  override def tryComplete(): Boolean = coordinator.tryCompleteJoin(group, forceComplete _)
-  override def onExpiration() = coordinator.onExpireJoin()
-  override def onComplete() = coordinator.onCompleteJoin(group)
+  /**
+   * It controls the lock manually since GroupCoordinator#onCompleteJoin() invoked by onComplete() can't be within a
+   * group lock since GroupCoordinator#onCompleteJoin() tries to complete delayed requests.
+   */
+  override def tryComplete(): Boolean = if (group.inLock(group.hasAllMembersJoined)) forceComplete() else false
```

Comment:
> So, we probably want to keep doing that.

It may produce deadlock if we hold the group lock for ```GroupCoordinator.onCompleteJoin```.  

```GroupCoordinator.onCompleteJoin``` is possible to append record to ```__consumer_offsets``` (see https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala#L1168), and hence it will try to complete other delayed joins which have groups belonging to same partition of ```__consumer_offsets```. 

That is why I make ```DelayedJoin``` control the group lock manually.  For another,  ```GroupCoordinator.onCompleteJoin``` is used by ```DelayedJoin``` only so it should be fine to change behavior of group lock in this case.

Reply:
> I am thinking that we could do the following.

Copy that.

Reply:
> So, we probably want to keep doing that.

hmmm, I misunderstand your point. Please ignore my first comment :)